doc/platform/nlapi/netlink_api_kernel_programming.md

```diff
@@ -0,0 +1,223 @@
+ 
+  
+# 3 Design  
+   
+## 3.1 Overview  
+Kernel programming using fork based iproute2 command call and newly proposed Netlink API call based Kernel programming is as shown below.
+  
+  
+![NLAPI](images/NLAPI_Block_Diagram.png "Figure 1: Kernel Programming current verses new design")  
+  
+__Figure 1: Kernel Programming current verses new design__  
+  
+ API function class will be introduced to serve specific set of command functionality. All these APIs are bundled into a library called nlapi. Applications needing to replace the exec calls can include this nlapi library and use the API calls.
+
+An example code snippet of the fork based call and Netlink API based call from SwSS VlanManager  is as below
+
+     popen fork based execution:
+     std::string res;
+     const std::string cmd = "/sbin/ip link add Bridge up type bridge" 
+     EXEC_WITH_ERROR_THROW(cmds, res); 
+       
+     Netlink API:
+     int ret = nlapi_req_ip_link_add_dev("Bridge", "bridge", "up");
+     NLAPI_STATUS_ERROR_THROW(ret);      
+
+
+### 3.1.1    Performance comparison
+
+In order to benchmark the performance improvement, SwSS Vlan manager was considered for measuring the performance difference between fork and API based execution. Significant performance improvement in SwSS Vlan configuration manager kernel programming was observed.  Measured data is as shown in table below. 
+
+Total time taken at SwSS: Shell fork calls vs Netlink API calls  
+
+| **Config Operation** | **Fork Calls** | **Netlink API Calls** | **Performance gain** |  
+|:---------------------|:--------------:|:---------------------:|:---------------------------:|  
+| One VLAN config when system is loaded     | 115 msec    | 08 msec    | 14 times |  
+| 4000 VLAN configuration                   | 305 seconds | 21 seconds | 14 times |  
+| 4000 VLAN config using VLAN range command | 309 seconds | 13 seconds | 23 times |  
+| 4000 VLAN config with system restart      | 406 seconds | 15 seconds | 27 times |  
+| 4000 VLAN config  with config file load   | 404 seconds | 17 seconds | 23 times |  
+
+### Table 2: Performance comparision  
+
+Command set used for above performance measurement.  
+
+       One Vlan config: 
+           config vlan add 9  
+           config vlan member add 9 Ethernet20  
+           config vlan member add 9 PortChannel005  
+       4000 Vlan config:   
+            config vlan range add 10 4009   
+            config vlan member range add 10 4009 Ethernet20  
+            config vlan member range add 10 4009 PortChannel005     
+
+
+Netlink API library will be placed in sonic-swss path with directory named 'nlapi'.
```

Comment:
Currently we have a Netlink message handler in sonic-swss-common. I think it will be better to have this nl library in swss-common as well.

Reply:
At present it is used only in SWSS, hence implemented in swss. Will look into if we can easily put it into swss common.